src/Components/Web.Extensions/src/InputFile/FileListEntry.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    internal class FileListEntry : IFileListEntry
+    {
+        private Stream? _stream;
+
+        internal InputFile Owner { get; set; } = default!;
+
+        public int Id { get; set; }
+
+        public string Name { get; set; } = string.Empty;
+
+        public DateTime? LastModified { get; set; }
+
+        public long Size { get; set; }
+
+        public string? Type { get; set; }
+
+        public string? RelativePath { get; set; }
+
+        public Stream Data
+        {
+            get
+            {
+                _stream ??= Owner.OpenFileStream(this);
+                return _stream;
+            }
+        }
+
+        public event EventHandler? OnDataRead;
```

Comment:
If we move to the `OpenFileStream` API, in which a given `IFileListEntry` might have any number of concurrent streams, then this `OnDataRead` API no longer really makes sense because it doesn't say which stream has read some data.

We could define our own `Stream` subclass, e.g., `FileUploadStream`, that exposes an `OnRead` event. However that's a big hammer of a solution that doesn't necessarily compose well with other ways of reading data.

On the other hand, we could decide that having an `OnRead` event is unnecessary altogether. It's only intended as a way of showing progress in the UI, so as an alternative, a developer could write their own logic to pull data from the stream in fixed-length increments and update some progress UI each time. It's really just about how focused we want to make the API on one specific common scenario (read a single stream and update a progress bar in the UI). My instinct would be to remove `OnDataRead` if we're unsure about it, because we can always add it in the future. 

What do you think? Is there any other design that comes to mind?

Reply:
I think if we were to have an `OnDataRead` event, the best way to do it (at least that I can think of at the moment) would be to make the existing `FileListEntryStream` public, and add an `OnDataRead` event there. Maybe this is slightly preferable to creating an entirely new `Stream` subclass, but it basically has the same effect on the public API surface.

I think it would be nice to have some sort of progress reporting system, as I would think this is could be a common scenario. However, there do some to be a few resources online describing various ways this can be done on a general case, so I think it's probably fine to leave out `OnDataRead` for now.

Reply:
If there are reasonably practical patterns for getting callbacks as the underlying stream progresses, without needing a separate `OnDataRead` event, then that seems much better. We can always add an event in the future, but can't take them away :)